Cluster deployment fails with 0.7.0-rc.1 and allowedCidrs set

/kind bug

Hello, im testing the `0.7.0-rc.1` release and im facing an issue deploying new clusters when allowedCidrs for the apiLoadbalancer are set.
Deploying a cluster without apiServerLoadbalancer.allowedCidrs set work just fine and i can add them after the deployment. This works. When i set the `allowedCidrs` from the start i get into a loop where FIPs are created and trying to be attached to the LB VIP while it already has FIP.


**What steps did you take and what happened:**
- Define `apiServerLoadbalancer.allowedCidrs`
- Deploy cluster
- Cluster never gets deployed as capo is stuck trying to attach FIPs to LB VIP (while it already has one attached). Some info/logs:

events from OpenStackCluster(openstack endpoint and actual ip is changed/dummied):
```
Events:
  Type     Reason                         Age    From                  Message
  ----     ------                         ----   ----                  -------
  Normal   Successfulcreatenetwork        5m58s  openstack-controller  Created network k8s-clusterapi-cluster-capi-dev-capi-dev with id e0f2a293-dfe8-4d60-90db-e6877c10cb14
  Normal   Successfulcreatesubnet         5m57s  openstack-controller  Created subnet k8s-clusterapi-cluster-capi-dev-capi-dev with id 12e4820c-f632-46bb-b3be-e1f5d632efcd
  Normal   Successfulcreaterouter         5m56s  openstack-controller  Created router k8s-clusterapi-cluster-capi-dev-capi-dev with id 9ac519b4-876f-4cf0-b301-d6953066be09
  Normal   Successfulcreatesecuritygroup  5m52s  openstack-controller  Created security group k8s-cluster-capi-dev-capi-dev-secgroup-controlplane with id 24e00e6d-a11e-4df4-a3eb-0ed3234107e4
  Normal   Successfulcreatesecuritygroup  5m52s  openstack-controller  Created security group k8s-cluster-capi-dev-capi-dev-secgroup-worker with id 56b5393e-ede4-4e9c-a8fe-6422ff0becb7
  Normal   Successfulcreatesecuritygroup  5m51s  openstack-controller  Created security group k8s-cluster-capi-dev-capi-dev-secgroup-bastion with id db87250d-c0f9-44ee-945a-42201f8444a2
  Normal   Successfulcreateloadbalancer   5m40s  openstack-controller  Created load balancer k8s-clusterapi-cluster-capi-dev-capi-dev-kubeapi with id 5176074d-c79f-4cc2-8428-9cd31f8f7a76
  Normal   Successfulcreatefloatingip     4m40s  openstack-controller  Created floating IP 8.8.8.122 with id 65dc92d6-505b-4ee8-81db-7b74b0d764cd
  Normal   Successfulassociatefloatingip  4m31s  openstack-controller  Associated floating IP 8.8.8.122 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db
  Normal   Successfulcreatelistener       4m27s  openstack-controller  Created listener k8s-clusterapi-cluster-capi-dev-capi-dev-kubeapi-6443 with id b9168acc-6e72-48ff-adb8-c6e68d12bf25
  Normal   Successfulcreatepool           4m25s  openstack-controller  Created pool k8s-clusterapi-cluster-capi-dev-capi-dev-kubeapi-6443 with id def2c36f-fad3-4787-bf65-a2d0c69223e4
  Normal   Successfulcreatemonitor        4m24s  openstack-controller  Created monitor k8s-clusterapi-cluster-capi-dev-capi-dev-kubeapi-6443 with id 802c09b2-4bc1-4233-bec5-c5137ac215ef
  Normal   Successfulcreatefloatingip     4m     openstack-controller  Created floating IP 8.8.8.30 with id 9afd400c-1c80-4f08-a1e0-e440dff6f562
  Warning  Failedassociatefloatingip      4m     openstack-controller  Failed to associate floating IP 8.8.8.30 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db: Expected HTTP response code [200] when accessing [PUT https://openstack-endpoint:9696/v2.0/floatingips/9afd400c-1c80-4f08-a1e0-e440dff6f562], but got 409 instead
{"NeutronError": {"type": "FloatingIPPortAlreadyAssociated", "message": "Cannot associate floating IP 8.8.8.30 (9afd400c-1c80-4f08-a1e0-e440dff6f562) with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db using fixed IP 10.6.0.117, as that fixed IP already has a floating IP on external network 5f62fb8a-2b58-4605-855d-a3e69e9de8ba.", "detail": ""}}
  Normal   Successfulcreatefloatingip  3m56s  openstack-controller  Created floating IP 8.8.8.227 with id c56aa4b2-6ef4-4561-95a7-838e210da10b
  Warning  Failedassociatefloatingip   3m56s  openstack-controller  Failed to associate floating IP 8.8.8.227 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db: Expected HTTP response code [200] when accessing [PUT https://openstack-endpoint:9696/v2.0/floatingips/c56aa4b2-6ef4-4561-95a7-838e210da10b], but got 409 instead
{"NeutronError": {"type": "FloatingIPPortAlreadyAssociated", "message": "Cannot associate floating IP 8.8.8.227 (c56aa4b2-6ef4-4561-95a7-838e210da10b) with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db using fixed IP 10.6.0.117, as that fixed IP already has a floating IP on external network 5f62fb8a-2b58-4605-855d-a3e69e9de8ba.", "detail": ""}}
  Normal   Successfulcreatefloatingip  3m52s  openstack-controller  Created floating IP 8.8.8.22 with id d0a052d4-c033-440d-b033-0ebf8ca7b18e
  Warning  Failedassociatefloatingip   3m51s  openstack-controller  Failed to associate floating IP 8.8.8.22 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db: Expected HTTP response code [200] when accessing [PUT https://openstack-endpoint:9696/v2.0/floatingips/d0a052d4-c033-440d-b033-0ebf8ca7b18e], but got 409 instead
{"NeutronError": {"type": "FloatingIPPortAlreadyAssociated", "message": "Cannot associate floating IP 8.8.8.22 (d0a052d4-c033-440d-b033-0ebf8ca7b18e) with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db using fixed IP 10.6.0.117, as that fixed IP already has a floating IP on external network 5f62fb8a-2b58-4605-855d-a3e69e9de8ba.", "detail": ""}}
  Normal   Successfulcreatefloatingip  3m48s  openstack-controller  Created floating IP 8.8.8.112 with id 319787ff-7852-438e-a58b-cccacb8a27ee
  Warning  Failedassociatefloatingip   3m47s  openstack-controller  Failed to associate floating IP 8.8.8.112 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db: Expected HTTP response code [200] when accessing [PUT https://openstack-endpoint:9696/v2.0/floatingips/319787ff-7852-438e-a58b-cccacb8a27ee], but got 409 instead
{"NeutronError": {"type": "FloatingIPPortAlreadyAssociated", "message": "Cannot associate floating IP 8.8.8.112 (319787ff-7852-438e-a58b-cccacb8a27ee) with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db using fixed IP 10.6.0.117, as that fixed IP already has a floating IP on external network 5f62fb8a-2b58-4605-855d-a3e69e9de8ba.", "detail": ""}}
  Normal   Successfulcreatefloatingip  3m44s  openstack-controller  Created floating IP 8.8.8.240 with id 95d86f64-d9de-42b2-93e3-9a26ae6552d2
  Warning  Failedassociatefloatingip   3m44s  openstack-controller  Failed to associate floating IP 8.8.8.240 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db: Expected HTTP response code [200] when accessing [PUT https://openstack-endpoint:9696/v2.0/floatingips/95d86f64-d9de-42b2-93e3-9a26ae6552d2], but got 409 instead
{"NeutronError": {"type": "FloatingIPPortAlreadyAssociated", "message": "Cannot associate floating IP 8.8.8.240 (95d86f64-d9de-42b2-93e3-9a26ae6552d2) with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db using fixed IP 10.6.0.117, as that fixed IP already has a floating IP on external network 5f62fb8a-2b58-4605-855d-a3e69e9de8ba.", "detail": ""}}
  Warning  Failedassociatefloatingip  3m40s  openstack-controller  Failed to associate floating IP 8.8.8.66 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db: Expected HTTP response code [200] when accessing [PUT https://openstack-endpoint:9696/v2.0/floatingips/40083e62-6bc5-45ca-80c0-2c82ebcb2c15], but got 409 instead
{"NeutronError": {"type": "FloatingIPPortAlreadyAssociated", "message": "Cannot associate floating IP 8.8.8.66 (40083e62-6bc5-45ca-80c0-2c82ebcb2c15) with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db using fixed IP 10.6.0.117, as that fixed IP already has a floating IP on external network 5f62fb8a-2b58-4605-855d-a3e69e9de8ba.", "detail": ""}}
  Normal   Successfulcreatefloatingip  3m40s  openstack-controller  Created floating IP 8.8.8.66 with id 40083e62-6bc5-45ca-80c0-2c82ebcb2c15
  Normal   Successfulcreatefloatingip  3m36s  openstack-controller  Created floating IP 8.8.8.56 with id 5cd30796-7a99-49ef-b8c4-2f587b941932
  Warning  Failedassociatefloatingip   3m36s  openstack-controller  Failed to associate floating IP 8.8.8.56 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db: Expected HTTP response code [200] when accessing [PUT https://openstack-endpoint:9696/v2.0/floatingips/5cd30796-7a99-49ef-b8c4-2f587b941932], but got 409 instead
{"NeutronError": {"type": "FloatingIPPortAlreadyAssociated", "message": "Cannot associate floating IP 8.8.8.56 (5cd30796-7a99-49ef-b8c4-2f587b941932) with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db using fixed IP 10.6.0.117, as that fixed IP already has a floating IP on external network 5f62fb8a-2b58-4605-855d-a3e69e9de8ba.", "detail": ""}}
  Normal   Successfulcreatefloatingip  3m31s  openstack-controller  Created floating IP 8.8.8.67 with id cb12e330-301f-4e96-9d0e-2415720c0f26
  Warning  Failedassociatefloatingip   3m31s  openstack-controller  Failed to associate floating IP 8.8.8.67 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db: Expected HTTP response code [200] when accessing [PUT https://openstack-endpoint:9696/v2.0/floatingips/cb12e330-301f-4e96-9d0e-2415720c0f26], but got 409 instead
{"NeutronError": {"type": "FloatingIPPortAlreadyAssociated", "message": "Cannot associate floating IP 8.8.8.67 (cb12e330-301f-4e96-9d0e-2415720c0f26) with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db using fixed IP 10.6.0.117, as that fixed IP already has a floating IP on external network 5f62fb8a-2b58-4605-855d-a3e69e9de8ba.", "detail": ""}}
  Normal   Successfulcreatefloatingip  3m27s  openstack-controller  Created floating IP 8.8.8.208 with id 9aa60ba0-0162-49a3-be79-84bcedcfb902
  Warning  Failedassociatefloatingip   3m27s  openstack-controller  Failed to associate floating IP 8.8.8.208 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db: Expected HTTP response code [200] when accessing [PUT https://openstack-endpoint:9696/v2.0/floatingips/9aa60ba0-0162-49a3-be79-84bcedcfb902], but got 409 instead
{"NeutronError": {"type": "FloatingIPPortAlreadyAssociated", "message": "Cannot associate floating IP 8.8.8.208 (9aa60ba0-0162-49a3-be79-84bcedcfb902) with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db using fixed IP 10.6.0.117, as that fixed IP already has a floating IP on external network 5f62fb8a-2b58-4605-855d-a3e69e9de8ba.", "detail": ""}}
  Warning  Failedassociatefloatingip  3m6s (x3 over 3m22s)  openstack-controller  (combined from similar events): Failed to associate floating IP 8.8.8.218 with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db: Expected HTTP response code [200] when accessing [PUT https://openstack-endpoint:9696/v2.0/floatingips/352c4a23-2456-48a6-835f-c015377f77a3], but got 409 instead
{"NeutronError": {"type": "FloatingIPPortAlreadyAssociated", "message": "Cannot associate floating IP 8.8.8.218 (352c4a23-2456-48a6-835f-c015377f77a3) with port 6e3faa7c-a1d5-43af-bd75-daa8b357a8db using fixed IP 10.6.0.117, as that fixed IP already has a floating IP on external network 5f62fb8a-2b58-4605-855d-a3e69e9de8ba.", "detail": ""}}
  Normal  Successfulcreatefloatingip  2m52s (x4 over 3m22s)  openstack-controller  (combined from similar events): Created floating IP 8.8.8.154 with id eb0c924f-b092-45ee-91b9-e062bec3ace0
```
Spec of OpenstackCluster(i use ClusterClass, so this is the rendered one):
```
spec:
  allowAllInClusterTraffic: false
  apiServerLoadBalancer:
    allowedCidrs:
    - 8.38.237.0/24
    - 8.100.1.0/24
    - 8.100.19.0/24
    - 8.90.232.0/23
    - 8.90.234.0/24
    - 8.100.43.7/32
    - 8.100.43.48/32
    - 8.100.43.54/32
    enabled: true
  bastion:
    enabled: true
    instance:
      cloudName: ""
      flavor: generic-2core-8gb-50gb
      image: ubuntu_22.04
      sshKeyName: capi-dev-ssh
  cloudName: capi-dev
  controlPlaneEndpoint:
    host: ""
    port: 0
  disableAPIServerFloatingIP: false
  dnsNameservers:
  - 8.100.29.131
  - 8.100.29.146
  externalNetworkId: 5f62fb8a-2b58-4605-855d-a3e69e9de8ba
  identityRef:
    kind: Secret
    name: capi-dev-cloud-config
  managedSecurityGroups: true
  network: {}
  nodeCidr: 10.6.0.0/24
  subnet: {}
  tags:
  - managed-by::cluster-api
  - cluster-name::capi-dev
```

Capo-logs:
[capo-logs.md](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/files/10391449/capo-logs.md)


**What did you expect to happen:**
- Cluster to get deployed as expected

**Anything else you would like to add:**
- **Im using ClusterClass**, the rendered OpenstackCluster seems just fine. I want to test it without clusterClass as well but im not sure ill have time these days
- Deploying the cluster without  `apiServerLoadbalancer.allowedCidrs` works just fine, after the deployment i can define it, apply and the loadbalancer listeners are updated without an issue. The issue occurs when i try to set them from the get-go.

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): 0.7.0-rc.1
- Cluster-API version: 1.3.1
- OpenStack version: Train
- Minikube/KIND version: 
- Kubernetes version (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

I don't have this env either, but from the info we got seems you are right, @bavarianbidi 
we should fix this issue as you proposed 

my question is about CI
1) can we update existing CI to cover this scenario (or maybe temply?)
2) I am curious on why the original error occurs .. if something wrong (nil point), then we will keep adding the floating ip?  this might be something need re-consider after nilpoint fixed

```
d0bc7360a101" type="Normal" object={Kind:OpenStackCluster Namespace:daphne Name:satellite-prod-01-88ll5 UID:8d6c312d-955a-49da-8a00-16e9e97ce7f2 APIVersion:infrastructure.cluster.x-k8s.io/v1alpha6 ResourceVersion:3115 FieldPath:} reason="Successfulcreatemonitor"
I0113 13:56:20.271017       1 controller.go:117] "Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference" controller="openstackcluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="OpenStackCluster" OpenStackCluster="daphne/satellite-prod-01-88ll5" namespace="daphne" name="satellite-prod-01-88ll5" reconcileID=690fb6c8-35bd-4fbc-9d31-4264da8854ce
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x108 pc=0x154c305]

goroutine 411 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:118 +0x1f4

```

--- a/pkg/cloud/services/loadbalancer/loadbalancer.go
+++ b/pkg/cloud/services/loadbalancer/loadbalancer.go
@@ -227,16 +227,26 @@ func (s *Service) getOrUpdateAllowedCIDRS(openStackCluster *infrav1.OpenStackClu
 	if len(openStackCluster.Spec.APIServerLoadBalancer.AllowedCIDRs) > 0 {
 		allowedCIDRs = append(allowedCIDRs, openStackCluster.Spec.APIServerLoadBalancer.AllowedCIDRs...)
 
-		if openStackCluster.Spec.Bastion.Enabled {
-			allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Bastion.FloatingIP, openStackCluster.Status.Bastion.IP)
+		// In the first reconciliation loop, only the Ready field is set in openStackCluster.Status
+		// All other fields are empty/nil
+		if openStackCluster.Status.Bastion != nil {
+			if openStackCluster.Status.Bastion.FloatingIP != "" {
+				allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Bastion.FloatingIP)
+			}
+
+			if openStackCluster.Status.Bastion.IP != "" {
+				allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Bastion.IP)
+			}
 		}
 
-		if openStackCluster.Status.Network.Subnet.CIDR != "" {
-			allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Network.Subnet.CIDR)
-		}
+		if openStackCluster.Status.Network != nil {
+			if openStackCluster.Status.Network.Subnet.CIDR != "" {
+				allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Network.Subnet.CIDR)
+			}
 
-		if len(openStackCluster.Status.Network.Router.IPs) > 0 {
-			allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Network.Router.IPs...)
+			if len(openStackCluster.Status.Network.Router.IPs) > 0 {
+				allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Network.Router.IPs...)
+			}
 		}
 	}
 

